crates/runtime/src/mmap.rs

```diff
@@ -44,6 +47,74 @@ impl Mmap {
         Self::accessible_reserved(rounded_size, rounded_size)
     }
 
+    /// Creates a new `Mmap` from the specified open `file` with the `len`
+    /// specified.
+    ///
+    /// The memory is mapped in read-only mode for the entire file. If portions
+    /// of the file need to be modified then the `region` crate can be use to
+    /// alter permissions of each page.
+    ///
+    /// Note that `len` doesn't have to be page-aligned, but rather it's
+    /// expected to be the length of the `file` provided.
+    pub fn new_file(file: File, len: usize) -> Result<Self> {
+        #[cfg(unix)]
+        {
+            use std::os::unix::prelude::*;
+            let ptr = unsafe {
+                libc::mmap(
+                    ptr::null_mut(),
+                    len,
+                    libc::PROT_READ,
+                    libc::MAP_PRIVATE,
+                    file.as_raw_fd(),
+                    0,
+                )
+            };
+            if ptr as isize == -1_isize {
+                return Err(io::Error::last_os_error())
+                    .context(format!("mmap failed to allocate {:#x} bytes", len));
+            }
+
+            Ok(Self {
+                ptr: ptr as usize,
+                len,
+                file: Some(file),
+            })
+        }
+        #[cfg(windows)]
+        {
+            use std::os::windows::prelude::*;
+            use winapi::um::handleapi::*;
+            use winapi::um::memoryapi::*;
+            use winapi::um::winnt::*;
+            unsafe {
+                let mapping = CreateFileMappingW(
+                    file.as_raw_handle().cast(),
+                    ptr::null_mut(),
+                    PAGE_READONLY,
+                    0,
+                    0,
+                    ptr::null(),
+                );
+                if mapping.is_null() {
+                    return Err(io::Error::last_os_error())
+                        .context(format!("failed to create file mapping of {:#x} bytes", len));
+                }
+                let ptr = MapViewOfFile(mapping, FILE_MAP_READ, 0, 0, len);
```

Comment:
I think we might need `FILE_MAP_COPY` here so that if we handle any relocations in the text section, the pages become private (and backed with the page file) rather than modifying the share pages which ultimately get flushed to the backed file (the precompiled module in this case).